src/functions.jl

```diff
@@ -0,0 +1,164 @@
+"""
+    map(f, ds, [key]; dims) -> KeyedDataset
+
+Apply function `f` to each component of the [`KeyedDataset`](@ref).
+Returns a new dataset with the same constraints, but new components.
+The function can be applied to a subselection of components via a [`Pattern`](@ref) `key` and/or `dims`.
+
+# Example
+```jldoctest
+julia> using AxisKeys, Statistics; using AxisSets: KeyedDataset;
+
+julia> ds = KeyedDataset(
+           :val1 => KeyedArray(zeros(3, 2); time=1:3, obj=[:a, :b]),
+           :val2 => KeyedArray(ones(3, 2); time=1:3, loc=[:x, :y]),
+       );
+
+julia> mean.((ds.val1, ds.val2))
+(0.0, 1.0)
+
+julia> r = map(a -> a .+ 1.0, ds; dims=:loc);
+
+julia> mean.((r.val1, r.val2))
+(0.0, 2.0)
+```
+"""
+function Base.map(f::Function, ds::KeyedDataset, key::Tuple; kwargs...)
+    return map(f, ds, Pattern(key); kwargs...)
+end
+
+function Base.map(f::Function, ds::KeyedDataset, key::Pattern=Pattern(:__); dims=:)
```

Comment:
Why support dims here if the convention is for only mapslices to support `dims`?

Reply:
It's so that you can apply a function to any component with those dimensions. Unlike `mapslices` this isn't applying the function to each slice along that `dim`.

Reply:
NOTE: That we actually use the `dims` argument in our `mapslices` wrapper to reduce the possible components to those that actually have those dimensions.

Reply:
Alternatively, we could merge the `key` and `dims` into `patterns::Pattern...` where the patterns must describe a `dimpath`?

Reply:
I am unclear on what `key` actually does here

Reply:
I like the idea of merging everything into `patterns`. Do we lose any functionality by doing that?

Reply:
> I am unclear on what key actually does here

I describes a set of `dimpath` constraints for selecting components to apply `f` too

> Do we lose any functionality by doing that?

No, you may just need to add an extra `:_` to the end of a pattern if you don't care about the dimension names of the component. I don't know how often we'll actually hit that case though.